### ISSUE TITLE
Ratchet & Clank: Tools of Destruction crashes on VK_ERROR_DEVICE_LOST

### ISSUE THREAD

**Issue by Valtekken** [RPCS3/rpcs3#12585]
### Quick summary

While playing Ratchet & Clank: Tools of Destruction, the game will randomly freeze without any sign of recovering whatsoever, forcing the user to shut down the emulator entirely or stop the emulation (which often causes the entire emulator to crash anyway).

### Details

As mentioned in the quick summary, the game freezes randomly and refuses to recover. The in game soundtrack will keep playing, but the visual part of the game (and the controls as well) becomes completely non-responsive. The one constant that seems to be there in this kind of issue is the fact that the log consistently shows an error such as "{RSX [0x211cd08]} SIG: Thread terminated due to fatal error: Assertion Failed! Vulkan API call failed with unrecoverable error: Device lost (Driver crashed with unspecified error or stopped responding and recovered) (VK_ERROR_DEVICE_LOST)" as the last piece of information before the freeze.

### Attach a log file

[TTY.log](https://github.com/RPCS3/rpcs3/files/9457465/TTY.log)
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9457466/RPCS3.log)

### Attach capture files for visual issues

Not needed

### System configuration

OS: Windows 11 21H2 (build 22000.918)
CPU: AMD Ryzen 5 5600X
GPU: NVidia GeForce GTX 970
GPU driver version: 516.94

### Other details

Emulator settings are stock, except for those recommended in the game's page on the wiki for optimal performance (though Async Texture Streaming was disabled to check if the problem could be mitigated that way, but no dice).

**Comment by YuriNator557** [comment]
These are definitely not default settings, reset them and try again

**Comment by Valtekken** [comment]
I know they're not default, I did mention in the final section that the only modified settings are the ones suggested by the wiki for this game

**Comment by YuriNator557** [comment]
Sorry I meant that a lot of these settings are not mentioned on the wiki

**Comment by Valtekken** [comment]
Okay then, I'll reset the settings and try again I guess.

**Comment by Valtekken** [comment]
Crashed again, default settings across the board.
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9461907/RPCS3.log)
[TTY.log](https://github.com/RPCS3/rpcs3/files/9461908/TTY.log)

**Comment by kd-11** [comment]
> Crashed again, default settings across the board. [RPCS3.log](https://github.com/RPCS3/rpcs3/files/9461907/RPCS3.log) [TTY.log](https://github.com/RPCS3/rpcs3/files/9461908/TTY.log)

There is no crash in this log.

**Comment by Valtekken** [comment]
> > Crashed again, default settings across the board. [RPCS3.log](https://github.com/RPCS3/rpcs3/files/9461907/RPCS3.log) [TTY.log](https://github.com/RPCS3/rpcs3/files/9461908/TTY.log)
> 
> There is no crash in this log.

Uhhh...no idea why, the screen froze just like last time. What am I supposed to do now?

**Comment by kd-11** [comment]
Depends on how easy it is to trigger the crash.
If you can crash it easily, install the vulkan SDK from [here](https://vulkan.lunarg.com/sdk/home#windows) and follow [these instructions.](https://github.com/RPCS3/rpcs3/issues/12084#issuecomment-1137799160)
Beware that enabling API debug will be very slow, so you need to pick an area where the crashing is easy to trigger.

**Comment by Valtekken** [comment]
> Depends on how easy it is to trigger the crash. If you can crash it easily, install the vulkan SDK from [here](https://vulkan.lunarg.com/sdk/home#windows) and follow [these instructions.](https://github.com/RPCS3/rpcs3/issues/12084#issuecomment-1137799160) Beware that enabling API debug will be very slow, so you need to pick an area where the crashing is easy to trigger.

I mean, the crashes happen sort of randomly in that for example the save I'm currently on is on an area of the game where I crashed literally 5 seconds after loading the level up, but also managed to play like a half hour without interruption, and every kind of length of time inbetween. I'll install that stuff and follow the instructions, but I have no idea whether I'll get the crash easily or not. Also, do I restore defaults again or do I just use the settings I used for a majority of the playthrough? I also cleaned all of the game's caches to see if that would help but had no luck in that sense.

**Comment by Darkhost1999** [comment]
Maintain defaults and keep in mind. People who get crashes all the time enable this debug output setting and can't recreate the issue. That's why finding a reliable method of creating the issue is important.

**Comment by Valtekken** [comment]
> Maintain defaults and keep in mind. People who get crashes all the time enable this debug output setting and can't recreate the issue. That's why finding a reliable method of creating the issue is important.

Okay then, I'm gonna restore defaults again and see where that takes me with the debug setting and the SDK, hopefully I'll be right back with another crash

**Comment by Valtekken** [comment]
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9463709/RPCS3.log)
Luckily, it crashed again ALMOST instantly. Performance didn't tank for some reason, maybe debug wasn't demanding enough?

**Comment by Valtekken** [comment]
Update: I suspected that the performance not tanking meant debug wasn't on, so I erased all settings from the start and just turned on debug again, and this time performance tanked AND I got a crash, so here's the file. Hopefully it helps.
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9463851/RPCS3.log)

**Comment by Valtekken** [comment]
> There are some messages, but nothing seems to be triggering the main crash. The only hint I have is that for some reason we consumed the entire command pool then immediately crashed afterward without any clear cause. I'll think of something to check what's going on.

Alright, I'll be waiting for further instructions then. Thanks for your work!

**Comment by kd-11** [comment]
I made some improvements recently to fix device lost in another game. Please check if it improved the situation here as well.

**Comment by Valtekken** [comment]
Alright, I'll make sure to check as soon as I get home. Thanks again!

**Comment by handle9** [comment]
Just a quick follow up on this. I'm seeing the same thing on a Ryzen 2600 and RX580. Everything runs fine then locks up. I'll figure out how to get logs later tonight.

**Comment by Valtekken** [comment]
> I made some improvements recently to fix device lost in another game. Please check if it improved the situation here as well.

Okay so, I'm a slight bit late compared to when I said I would check but I got around to it, and I got updates (though their usefulness is limited for sure): the game is certainly far more stable now, I was able to progress further without any crashes but at some point the crash happened once again. All of this was with my custom settings though, so as I said it has limited utility. Guess I'll try again with default settings and the debug option turned on at a later date to provide further data

**Comment by kd-11** [comment]
Any update?

**Comment by Valtekken** [comment]
> Any update?

Yes, a recent one. I played through a few sections that used to trigger freezes before this, using the latest build as of two days ago, and no freezes at all happened except at the very end, when I was already getting ready to shut things down and go to bed. Things have definitely improved it seems, but there's still an issue that's causing the freezes somewhere. Maybe driver updates, OS updates and emulator updates are mitigating it though. I will also say I enabled PBO on my CPU so that provided some extra juice, maybe

**Comment by kd-11** [comment]
I'll need a logs from any recent crashes as well.

**Comment by Valtekken** [comment]
> I'll need the logs from any recent crashes as well.

Yes, of course. I suppose the logs get archived, right? If they don't, I'll have to make another crash happen because I have opened RPCS3 to update it multiple times since the last crash. If they do, I guess I have to ask where RPCS3 stores them

**Comment by thatmaceguy** [comment]
Running into the same problem. Figured I'd tag onto this instead of starting a new issue.

Windows 10 Home 64-bit Build 19045
AMD Ryzen 7 1700
NVidia GeForce GTX970 528.49 (clean install prior to latest crash, tried last 4 versions)
Vulkan SDK (1.3.239.0) installed
Booting the game with global config reset to defaults.

RPCS3 crashes when I enable Debug Output. See RPCS3 log below. TTY log was empty. 
[RPCS3 202302241425.log](https://drive.google.com/file/d/14kbKNpU91nsKg910QA2pr25xspBG8pTO/view?usp=sharing)

The game will boot and run for various amount of time before crashing with VK_ERROR_DEVICE_LOST. Here are the RPCS3 and TTY logs, in case they are any help without debugging output enabled. 
[RPCS3 202302241424.log](https://drive.google.com/file/d/14o15f-oHtrUyZZHNVdKoJikVa9AAlxX2/view?usp=sharing)
[TTY 202302241424.log](https://drive.google.com/file/d/14rjP-mwGs7Lt06dJVSFNWaeAIy_8zOt_/view?usp=sharing)

I tried booting the game a few times with debugging output enabled, always the same crash. Turned it off, game boots fine, played for a few minutes then crashed (log above w/o debugging). Turned debugging output back on and tried booting the game again, crashed (log above w/debug on)

Let me know what else I can try/provide.

**Comment by thatmaceguy** [comment]
After reinstalling the Vulkan SDK (1.3.239.0) with all optional components and installing the Vulkan Runtime (same ver.), RPCS3 was finally able to boot R&CToD with Debug Output enabled. I played a section that reliably crashed (Clank in Iris Supercomputer) and, weirdly, made it all the way through before Vulkan finally crashed. Logs are below. 

[RPCS3 202302271723.log](https://drive.google.com/file/d/14yj7PkJB5ffsgFiwUhoT2Y5z3wtn57mD/view?usp=sharing)
[TTY 202302271723.log](https://drive.google.com/file/d/154GjwCWk3WcW_bh6hPkDN6V-gvMoNIR5/view?usp=sharing)

Hopefully this helps.

**Comment by thatmaceguy** [comment]
New nVidia drivers rolled out today, figured I'd try this again. 

RPCS3 v0.0.27-14800-0d34b979 Alpha
nVidia 531.18 (clean install)

Ran smoothly through the Kreeli Comet (all prior Vulkan crashes had been in the Iris Supercomputer section) and well into the next planet (Viceron) before finally crashing again just when I was beginning to think it was solved. 

Unfortunately, enabling Debug Output once again causes RPCS3 to crash while loading the game, though not quite as immediately as before. I'll try reinstalling the Vulkan SDK again tomorrow and see if that somehow clears it up. Till then, here are the non-debug logs (ran twice) in case anything useful can be found in them. 

[RPCS3 202303042102.log](https://drive.google.com/file/d/15kE6bKJahXxKhslnoQWYiPLKD0hdn_jB/view?usp=sharing)
[RPCS3 202303042104.log](https://drive.google.com/file/d/15jgYOJe4ywSo7ZX0nQY2d_UFDWVL2TV-/view?usp=sharing)

**Comment by Valtekken** [comment]
I sorta have a hunch the infamous VRAM config of the 970 may somehow be behind this. Many of the issues that mention this error report a system configuration including the 970, there must be a reason for that.

**Comment by thatmaceguy** [comment]
@Valtekken devs might have better luck if you submitted new crash logs too.

**Comment by Valtekken** [comment]
> @Valtekken devs might have better luck if you submitted new crash logs too.
I would, but I asked whether I needed to upload previous logs since the last ones I uploaded or provide entirely new logs, and where I could find the previous logs if the former, but no answer.

**Comment by thatmaceguy** [comment]
@Valtekken I mean, they're volunteers on an open source project. Probably thought you could figure it out on your own... Not their job to hold your hand. 

Best I can tell, logs are overwritten each time you fire up RPCS3. I just make a copy after each crash. Doesn't matter anyway, because there are new nVidia drivers out and a new release of RPCS3. Between the two I got through a section that was guaranteed to crash. Still crashed later on, but something clearly changed.

Right now, I stuck with RPCS3 itself crashing with Debug Output enabled. Hoping @kd-11 has some suggestions there.

**Comment by Valtekken** [comment]
> @Valtekken I mean, they're volunteers on an open source project. Probably thought you could figure it out on your own... Not their job to hold your hand.
> 
> Best I can tell, logs are overwritten each time you fire up RPCS3. I just make a copy after each crash. Doesn't matter anyway, because there are new nVidia drivers out and a new release of RPCS3. Between the two I got through a section that was guaranteed to crash. Still crashed later on, but something clearly changed.
> 
> Right now, I stuck with RPCS3 itself crashing with Debug Output enabled. Hoping @kd-11 has some suggestions there.

Yeah, and it's not like I get paid to report issues and upload logs either. This isn't about getting paid, I'm just a random end user and I can't be expected to figure out this more technical stuff. I know my way around a PC, most definitely not around bug reports and the like. I need help to figure out this stuff.

Having said that, no way in hell RPCS3 doesn't store past logs. And since kd didn't clarify what logs were needed, I erred on the side of caution and asked where I could get the previous logs. They could contain useful info.

I still suspect it's the stupid 3.5+0.5 VRAM's fault.

**Comment by kd-11** [comment]
Try with https://github.com/RPCS3/rpcs3/pull/13573

**Comment by MSuih** [comment]
https://wiki.rpcs3.net/index.php?title=Help:Using_different_versions_of_RPCS3#How_to_download_builds_from_a_pull_request_on_GitHub?

**Comment by Valtekken** [comment]
> [Follow the instructions here](https://wiki.rpcs3.net/index.php?title=Help:Using_different_versions_of_RPCS3#How_to_download_builds_from_a_pull_request_on_GitHub?)

Thanks!

**Comment by Valtekken** [comment]
> Try with #13573

Sadly, the game still crashes. I'm now at a point where I can reliably expect it to crash during any arena challenge on Mukow. Must be something else.

**Comment by kd-11** [comment]
Is this happening to anyone with a more recent GPU? This may be a driver branch thing.
In fact is it only happening with GTX 970s? If so I can add a workaround to limit VRAM usage to 3.5G max.

**Comment by kd-11** [comment]
I'm cooking up a special build to test if this is caused by the 970's memory partition issues.

**Comment by Valtekken** [comment]
> I'm cooking up a special build to test if this is caused by the 970's memory partition issues.

Thank you, I wanted to test this theory ever since the idea popped up in my head. If this fixes the issue it's yet another instance of NVidia's awful decisions having consequences on its customers years down the line.

**Comment by kd-11** [comment]
I created a draft PR to play around with. https://github.com/RPCS3/rpcs3/pull/13577
Also keep an eye on the task manager GPU memory tab, it may catch the moment you see a crash. The R&C engine can sometimes have intense memory utilization even on other 4GB cards due to how it allocates VRAM, basically we cannot react fast enough some times and it spams requests at the emulator.

**Comment by Valtekken** [comment]
> I created a draft PR to play around with. #13577 Also keep an eye on the task manager GPU memory tab, it may catch the moment you see a crash. The R&C engine can sometimes have intense memory utilization even on other 4GB cards due to how it allocates VRAM, basically we cannot react fast enough some times and it spams requests at the emulator.

Reporting back. It lasted a LOT longer than the previous tries before crashing in the arena (the particular challenge I'm grinding on lasts 2 exact minutes and I managed to run out of time on roughly 4 or 5 tries, whereas before it crashed after exactly 1 try, add 20 seconds to get to the arena via taxi and 5 seconds to select the challenge for both builds, stable and draft). I checked the Task Manager's VRAM graph and it seems to have kept stable for a long time at 1.5GBs of VRAM. Any ideas?

**Comment by Valtekken** [comment]
Also, after switching tabs to the Task Manager, the image that RPCS3 is not the one the game crashed on. The screen shows Ratchet inside the taxi, before getting into the arena. Image for reference.
![image](https://user-images.githubusercontent.com/112526564/227530737-c421985d-09fc-497f-858f-decda5c41a76.png)

**Comment by kd-11** [comment]
We have some aggressive reclamation in place that will automatically clean up memory allocations. Attach the log file, hopefully we can see something useful from the memory manager.

**Comment by kd-11** [comment]
> Also, after switching tabs to the Task Manager, the image that RPCS3 is not the one the game crashed on. The screen shows Ratchet inside the taxi, before getting into the arena. Image for reference. ![image](https://user-images.githubusercontent.com/112526564/227530737-c421985d-09fc-497f-858f-decda5c41a76.png)

This is "fine" - basically since the driver switched to exclusive fullscreen when you went into fullscreen mode the compositor has no idea what the actual screen looks like. It only has a snapshot from the last time it was activated.

**Comment by Valtekken** [comment]
> We have some aggressive reclamation in place that will automatically clean up memory allocations. Attach the log file, hopefully we can see something useful from the memory manager.

Alright, here it is.

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11062419/RPCS3.log)

**Comment by Valtekken** [comment]
> > Also, after switching tabs to the Task Manager, the image that RPCS3 is not the one the game crashed on. The screen shows Ratchet inside the taxi, before getting into the arena. Image for reference. ![image](https://user-images.githubusercontent.com/112526564/227530737-c421985d-09fc-497f-858f-decda5c41a76.png)
> 
> This is "fine" - basically since the driver switched to exclusive fullscreen when you went into fullscreen mode the compositor has no idea what the actual screen looks like. It only has a snapshot from the last time it was activated.

Oh, okay. Had no idea, learned something new.

**Comment by kd-11** [comment]
I lowered the limit some more, check if there is any difference. I'm basically just looking for any feedback loop that would confirm if we need to keep looking at memory behaviour or if we're looking in the wrong place completely.

**Comment by Valtekken** [comment]
> I lowered the limit some more, check if there is any difference. I'm basically just looking for any feedback loop that would confirm if we need to keep looking at memory behaviour or if we're looking in the wrong place completely.

Okay, will try the new build and report back ASAP.

I got another idea too as to where we could be looking at if it isn't memory, but I'll preface it by saying I'm very much not on the deep technical side of these things so this is just observation and personal experience relating to similar issues. I have no idea whether I'm saying some incredibly amateur, absolute buffoon shit or if I have any kind of correct hint about where the issue could be.

I also play games on Xemu, the Xbox emulator, and I've fairly recently had an issue where the framerate of the game I was playing tanked massively for roughly 5 minutes. After 5 minutes the framerate returned back to normal levels for the game/system config, and it stayed that way for another 5 minutes, then it tanked again and stayed incredibly low for 5 minutes. Basically it was a sort of green light-red light with framerate. After some investigating, turns out that one of the NVidia optimizations on multi-threaded workloads was causing the issues. My idea, having said this, is to time the crashes on a range of say, 10 sessions, and see if they're consistent in any way. If they are, maybe it's a similar issue with a similar origin.

As I said, this isn't coming from any place of actual deep technical knowledge about the inner workings of GPUs or emulation so disregard all of this if it isn't plausible.

**Comment by Valtekken** [comment]
> I lowered the limit some more, check if there is any difference. I'm basically just looking for any feedback loop that would confirm if we need to keep looking at memory behaviour or if we're looking in the wrong place completely.

Reporting back. Again, a crash, but the stable state seems to have lasted even longer, maybe 2 or 3 minutes, before freezing. Audio still playing as usual. GPU VRAM usage was 2.4 GBs at freeze time.

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11071487/RPCS3.log)

**Comment by kd-11** [comment]
That's actually relevant, since I locked the vram to 2.5G. I think maybe the system is crashing when doing some cleanup.

**Comment by Valtekken** [comment]
> That's actually relevant, since I locked the vram to 2.5G. I think maybe the system is crashing when doing some cleanup.

Is there anything I can do about that?

**Comment by thatmaceguy** [comment]
> I created a draft PR to play around with. #13577 Also keep an eye on the task manager GPU memory tab, it may catch the moment you see a crash. The R&C engine can sometimes have intense memory utilization even on other 4GB cards due to how it allocates VRAM, basically we cannot react fast enough some times and it spams requests at the emulator.

Ran with the draft PR about 30 minutes today with no major issues (just a few stutters). Will test more thoroughly later and monitor gpu vram usage.

**Comment by thatmaceguy** [comment]
The game ran well with #13577 on both default and custom settings well past where I had been getting consistent crashes. Made it about 20 minutes on custom settings and about 10 minutes on default settings immediately after. VRAM usage would cruise at about 60-65% , then jump to 75%, then crash soon after (in both configurations). 

I was also able to run with Debug Output enabled on this test build (previously, RPCS3 would crash as soon as I booted the game with this enabled). However, in debug the game never crashed despite VRAM usage reaching 78%. After about an hour of playing at 4-15fps without a crash, my eyes couldn't take it anymore so I shut it down. 

Let me know if you'd like to me try anything else. 

[RPCS3 202303261533 custom settings.log](https://drive.google.com/file/d/15wX2UDM5rQ6dJEuPKs8PrWCROm-F-OZ8/view?usp=sharing)
[TTY 202303261533 custom settings.log](https://drive.google.com/file/d/162RpA1vnAxBgk5NGQkNbN25uout2qb1M/view?usp=sharing)

[RPCS3 202303261548 default settings.log](https://drive.google.com/file/d/16GbfUzLSaiuu83WLYPHGds2cHRkZSqOb/view?usp=sharing)
[TTY 202303261548 default settings.log](https://drive.google.com/file/d/16FUFZ7u6e2hoUgxfDyNcJ5TbVmR0gfvL/view?usp=sharing)

[RPCS3 202303261652 debug no crash.log](https://drive.google.com/file/d/16INM4O9S5XGv7xYfOH0mD-b-Lzr7E7dJ/view?usp=sharing)
[TTY 202303261652 debug no crash.log](https://drive.google.com/file/d/16N2eySKKekCCS7PcMsyPDoRqimcWda04/view?usp=sharing)

**Comment by kd-11** [comment]
I guess you can try disabling vulkan memory allocator in debug settings to let the nvidia driver do whatever it wants.
Performance will be terrible, but maybe it won't crash.

**Comment by ojt1998** [comment]
I have a gtx1050 and also having the same crash. Currently trying with the vulkan memory allocator disabled

**Comment by kd-11** [comment]
I ran an experiment where I locked my reported VRAM to be low (512M) and using MSAA + upscaling I was able to crash the game with device lost on NVIDIA within a few seconds. It seems the hunch about VRAM being the trigger is correct.

**Comment by Valtekken** [comment]
> I ran an experiment where I locked my reported VRAM to be low (512M) and using MSAA + upscaling I was able to crash the game with device lost on NVIDIA within a few seconds. It seems the hunch about VRAM being the trigger is correct.

I'm feeling really proud of myself right now, not gonna lie.

**Comment by MineSAT** [comment]
> Ejecuté un experimento en el que bloqueé mi VRAM informado para que fuera bajo (512 M) y usando MSAA + mejora pude bloquear el juego con el dispositivo perdido en NVIDIA en unos pocos segundos. Parece que la corazonada de que la VRAM es el desencadenante es correcta.

do you think the vram memory usage limit could be increased to avoid that?

**Comment by Valtekken** [comment]
> do you think the vram memory usage limit could be increased to avoid that?

The issue probably isn't the limit itself, it's making the emulator work around the low amount of VRAM there is (if this is at all possible)

**Comment by kd-11** [comment]
The limit I added isn't a hard resource limit, just a weight for some routines. It doesn't really force the memory to never go beyond, there's a separate way to do that.
With that said, I'm pretty sure I figured out why NV driver is crashing when memory runs out. Once the fix is done, I'll remove the cap and allow 970s to use more memory.

**Comment by Valtekken** [comment]
> The limit I added isn't a hard resource limit, just a weight for some routines. It doesn't really force the memory to never go beyond, there's a separate way to do that. With that said, I'm pretty sure I figured out why NV driver is crashing when memory runs out. Once the fix is done, I'll remove the cap and allow 970s to use more memory.

I might be asking this too soon, but could you quickly explain where the issue lies exactly?

**Comment by kd-11** [comment]
It's a little complicated but basically we have the GPU holding a descriptor reference (this is normal) then we run out of vram for whatever reason in the middle of a draw call and remove some textures/buffers/whatever but the reference remains. When the system recovers and tries to send the next draw, we'll have shaders loaded that use the old reference to something that doesn't exist anymore. so GPU hard crashes. It's also why I was never able to reproduce the issue, you need hardware with low enough VRAM. The R&C engine leaks surfaces a lot and we don't clean up aggressively enough for performance reasons, so 4GB is nowhere near enough. Lowering the limit helped a little because the regular cleanup panicked that we're running low and cleans things up a little better.

**Comment by Valtekken** [comment]
> It's a little complicated but basically we have the GPU holding a descriptor reference (this is normal) then we run out of vram for whatever reason in the middle of a draw call and remove some textures/buffers/whatever but the reference remains. When the system recovers and tries to send the next draw, we'll have shaders loaded that use the old reference to something that doesn't exist anymore. so GPU hard crashes. It's also why I was never able to reproduce the issue, you need hardware with low enough VRAM. The R&C engine leaks surfaces a lot and we don't clean up aggressively enough for performance reasons, so 4GB is nowhere near enough. Lowering the limit helped a little because the regular cleanup panicked that we're running low and cleans things up a little better.

Perfect, now I get what the issue was. Is the fix particularly hard to design or implement?

**Comment by Valtekken** [comment]
> Hard to diagnose. It's a hydra. I kill one bug and 2 more appear somewhere else. That's because memory usage is not centralized (something to look at in future). In the end I have to compile my own vulkan validation tools with modifications to isolate the issue which is a massive waste of time. It'll be a few more days before I can get all the corner cases. As things are right now it's pretty good, I have to really push things to get crashes to happen but I don't like the fact that it still occurs.

I mean hey, if I get a crash less often it's still an improvement.

Regardless of how long it takes, thank you for your work. As far as I'm concerned R&CToD will be perfect with this fix and the text fix, which is already an astounding achievement on your part and on the team's part.

**Comment by kd-11** [comment]
Try https://github.com/RPCS3/rpcs3/pull/13859

**Comment by Valtekken** [comment]
> Try #13859

Sorry to be the bearer of bad news, but it still crashed as usual. Lasted half an hour this time, and to be fair it WAS one of the most intensive areas of the game...but yeah, still crashed.

**Comment by kd-11** [comment]
> > Try #13859
> 
> Sorry to be the bearer of bad news, but it still crashed as usual. Lasted half an hour this time, and to be fair it WAS one of the most intensive areas of the game...but yeah, still crashed.

Log?

**Comment by Valtekken** [comment]
> > > Try #13859
> > 
> > 
> > Sorry to be the bearer of bad news, but it still crashed as usual. Lasted half an hour this time, and to be fair it WAS one of the most intensive areas of the game...but yeah, still crashed.
> 
> Log?

Says file is too big, had to compress: [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/11473025/RPCS3.zip)

**Comment by kd-11** [comment]
Seems to be as expected. 4k + MSAA = 8k * 2 internally (MSAA images are shadowed for technical reasons).
You can probably "fix" the issue by lowering the scaling to 150% or thereabout. I'll be adding more fixes to the memory management to fix the issue once and for all at a later date.

**Comment by Valtekken** [comment]
> Seems to be as expected. 4k + MSAA = 8k * 2 internally (MSAA images are shadowed for technical reasons). You can probably "fix" the issue by lowering the scaling to 150% or thereabout. I'll be adding more fixes to the memory management to fix the issue once and for all at a later date.

Alright, I'll be waiting for the definitive fix then (and for the text fix of course)

Thanks for the help

**Comment by kd-11** [comment]
For the record, even when fixed, 4K on this game engine and a 970 may not be doable without severe VRAM-related slowdown or outright crashing, but the GPU driver shouldn't crash. Just setting expectations. The R&C future engine gobbles up memory like crazy when mapped onto a PC GPU.

**Comment by Valtekken** [comment]
> For the record, even when fixed, 4K on this game engine and a 970 may not be doable without severe VRAM-related slowdown or outright crashing, but the GPU driver shouldn't crash. Just setting expectations. The R&C future engine gobbles up memory like crazy when mapped onto a PC GPU.

I'll be honest, I don't even want 4K because my screen is 1080p, but if I go anywhere below that the graphics don't look like the resolution is set to 1080p. I do 4K to get 1080p, and if I had to actually get 4K I think I'd need to set it to 16K or some shit lmao

But yeah, guess I'll be upgrading sooner or later. Whenever a 60/600 tier GPU reaches 200€...

**Comment by Valtekken** [comment]
> @Valtekken Our wiki recommends people don't use MSAA for Tools of Destruction because it makes the game look pixelated, have you tried turning MSAA off? https://wiki.rpcs3.net/index.php?title=Ratchet_%26_Clank_Future:_Tools_of_Destruction

My custom config screen shows anti aliasing is disabled, unless you mean "Anisotropic filter" which is set at 16x (I usually slide it all the way to the top in games)

**Comment by Valtekken** [comment]
> Seems to be as expected. 4k + MSAA = 8k * 2 internally (MSAA images are shadowed for technical reasons). You can probably "fix" the issue by lowering the scaling to 150% or thereabout. I'll be adding more fixes to the memory management to fix the issue once and for all at a later date.

I retested it on the newest build with 150% scaling, output scaling set to nearest, anisotropic filtering set to 16x and antialiasing disabled (this last one, as I said before, was already disabled, but I'm confirming it again). Graphics look worse, as expected, but the game still crashed - any ideas? Maybe it's the anisotropic filtering?

**Comment by kd-11** [comment]
It could also be some other unrelated issue. I'll have to check later.
You can add a VRAM cap manually now if you want to test, check the config.yml file. You can set any number in megabytes (e.g 3000 will give you around 3 gigabytes of VRAM working pool)

**Comment by Valtekken** [comment]
> It could also be some other unrelated issue. I'll have to check later. You can add a VRAM cap manually now if you want to test, check the config.yml file. You can set any number in megabytes (e.g 3000 will give you around 3 gigabytes of VRAM working pool)

Okay, I'll try going with 250MB increments then, see where I get the best and worst results. Will report back when I'm done.

**Comment by kd-11** [comment]
> Terribly sorry to bother, but I can't seem to find the cap setting in the config.yml file. What's it named?

VRAM allocation limit

**Comment by Megamouse** [comment]
you probably have to save the config once to see it

**Comment by Valtekken** [comment]
> you probably have to save the config once to see it

As in saving the file in a text editor or as in saving the config from RPCS3's UI?

**Comment by kd-11** [comment]
> > you probably have to save the config once to see it
> 
> As in saving the file in a text editor or as in saving the config from RPCS3's UI?

Click on the game's custom config in the UI and save then search for the option and modify.

**Comment by kd-11** [comment]
For this game it seems there is something else that's broken unrelated to VRAM (I can reproduce with very low res scale and 12GB VRAM). There are multiple bugs at play here, we just have to get them one at a time.

**Comment by Valtekken** [comment]
> For this game it seems there is something else that's broken unrelated to VRAM (I can reproduce with very low res scale and 12GB VRAM). There are multiple bugs at play here, we just have to get them one at a time.

So maybe not ALL of the crashes are related to the VRAM, right? Or maybe none of them have anything to do with the VRAM and it's something else entirely?

**Comment by Darkhost1999** [comment]
At this time we're really just waiting for Nvidia to implement that Vulkan device fault extension right? Cause then we could obtain debug logs where Nvidia would be telling us what's at fault when it crashes is that correct?

**Comment by Valtekken** [comment]
> Maybe, but first you need to disable the native driver memory allocator (in the debug tab). I have isolated one problem in the 'read depth buffers' option, but that may still not be the same one you have. It really could be anything, only nvidia themselves can tell us for certain why some crashes happen.

Last time I disabled it, before the current builds with the newest VRAM fix, the game crashed anyway. Should I retry?

**Comment by kd-11** [comment]
> At this time we're really just waiting for Nvidia to implement that Vulkan device fault extension right? Cause then we could obtain debug logs where Nvidia would be telling us what's at fault when it crashes is that correct?

That's the end goal, though it is an optional extension, it may never come to fruition.

> Last time I disabled it, before the current builds with the newest VRAM fix, the game crashed anyway. Should I retry?

For testing never use the driver's raw allocator. That is guaranteed to always cause other issues.

**Comment by Valtekken** [comment]
> For testing never use the driver's raw allocator. That is guaranteed to always cause other issues.

I didn't use it on a regular basis, just the one time you told that other guy to use that debug option.

**Comment by Valtekken** [comment]
> I didn't use it on a regular basis, just the one time you told that other guy to use that debug option.

Apparently I forgot to turn that off so it was on during all these tests. The bad news is that by turning it off the crashes happened EVEN FASTER. We're talking like, within 5 minutes from starting the game, without any changes to the config file. Will be trying with changes to the config file now.

**Comment by Valtekken** [comment]
Test results:
4000=35 secs
3750=22mins30secs
3500=3mins40secs (this one failed with "fault in uninterruptible code")
3250=1min30secs
3000=4min30secs

There doesn't seem to be any pattern, sadly.

**Comment by Valtekken** [comment]
> I finally found the root cause. I'll open the PR soon. While one set of crashes was caused by VRAM the other one has nothing to do with it and is caused by some bad clipping math.

Man, I'll be making you a statue by the end of this ordeal lmao

**Comment by Valtekken** [comment]
Sad to report the game still crashes. Tried twice with 300% scaling, 16x anisotropic filter, bilinear method, crashed pretty quickly both times. Tried once with 150% scaling, auto anisotropic filter, nearest method, crashed like 20 seconds later than it would've with the 300% scaling based on the last two tries. Additional details: the VRAM limit is set to 4000.
Log: [RPCS3.log](https://github.com/RPCS3/rpcs3/files/11500881/RPCS3.log)

**Comment by kd-11** [comment]
I can't reproduce any more crashes on my system unfortunately. I played through the initial level several times with the current build which was impossible before.

**Comment by Valtekken** [comment]
> I can't reproduce any more crashes on my system unfortunately. I played through the initial level several times with the current build which was impossible before.

I see the new build has been changed since I tested it, should I download the new artifact and test again?

**Comment by kd-11** [comment]
I don't think it will fix your issue. The first PR build already had the core changes that made the crash go away on Ampere.
There are probably many different crashes affecting different GPUs differently here.
Try setting "strict rendering mode" to true btw just in case some speedhack is causing crashes on maxwell.

**Comment by Valtekken** [comment]
> I don't think it will fix your issue. The first PR build already had the core changes that made the crash go away on Ampere. There are probably many different crashes affecting different GPUs differently here. Try setting "strict rendering mode" to true btw just in case some speedhack is causing crashes on maxwell.

Do you want my save? This part has a reliable crash, just fly around from the first platform you see and it should crash, if it doesn't then NVidia screwed me over big time with Maxwell's drivers (especially considering support for that architecture is almost done, I think...it's been 9 years basically)
But I will try setting strict rendering mode to true, maybe that fixes it

**Comment by kd-11** [comment]
If it doesn't work attach your save and I'll try to reproduce. If I can cause the crash locally I can fix it pretty quickly now.

**Comment by Valtekken** [comment]
> If it doesn't work attach your save and I'll try to reproduce. If I can cause the crash locally I can fix it pretty quickly now.

I'm currently still testing, but it seems stable enough. No crashes on a 30min run. On the other hand, performance is absolutely atrocious AND graphics look awful, so maybe it's managing to hold because of that

**Comment by Valtekken** [comment]
Alright, it crashed EVEN with the strict rendering mode. Good news is that it didn't crash with VK_ERROR_DEVICE_LOST based on what I see from the log.

Log: [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/11502421/RPCS3.zip)
Savegame to reproduce crashes: [BCES00052_SAVE_1.zip](https://github.com/RPCS3/rpcs3/files/11502425/BCES00052_SAVE_1.zip)

Even if this doesn't get fixed, it's alright. If you feel that your energy would be better spent on other issues, please do. R&C ToD itself has other bugs of its own, like the text glitch, the pixelated water or the performance tanking when the Pyro Blaster is used, so there's plenty to do even on this game alone.

**Comment by papaenojado23** [comment]
Dear all, I'm just playing this game, I'm near the end, and I also had a lot of crashes, the game stays like a photo and you can't do anything other than close it

**Comment by kd-11** [comment]
> Dear all, I'm just playing this game, I'm near the end, and I also had a lot of crashes, the game stays like a photo and you can't do anything other than close it

That's a different issue. That's a lockup/freeze and not a GPU driver crash. You can open a separate ticket for that.

**Comment by kd-11** [comment]
> Alright, it crashed EVEN with the strict rendering mode. Good news is that it didn't crash with VK_ERROR_DEVICE_LOST based on what I see from the log.
> 
> Log: [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/11502421/RPCS3.zip) Savegame to reproduce crashes: [BCES00052_SAVE_1.zip](https://github.com/RPCS3/rpcs3/files/11502425/BCES00052_SAVE_1.zip)
> 
> Even if this doesn't get fixed, it's alright. If you feel that your energy would be better spent on other issues, please do. R&C ToD itself has other bugs of its own, like the text glitch, the pixelated water or the performance tanking when the Pyro Blaster is used, so there's plenty to do even on this game alone.

From the log it appears the game never crashed. You just experienced the lockup bug which is non-graphical in domain.
I cannot reproduce the crash, I tried flying around for about a minute or two and nothing happened. I'll create a build with some optimizations disabled, maybe that fixes the problem.
I also need to make sure the feature set matches since I'm sure there's a bunch of extensions not present on the 970.

**Comment by kd-11** [comment]
The freeze should be another ticket.
Now as for closing this ticket I still see one report with a device lost message and that needs investigation.

**Comment by Valtekken** [comment]
> The freeze should be another ticket.
> Now as for closing this ticket I still see one report with a device lost message and that needs investigation.

Alright then, I'll leave this one here and open a new one!

**Comment by kd-11** [comment]
Test this build once its done building in like 30 minutes
https://github.com/kd-11/rpcs3/runs/13593717063
Disables some NVIDIA speedhacks entirely. They're still enabled on pre-pascal GPUs and are known to cause crashes on RTX series.

**Comment by Valtekken** [comment]
> Test this build once its done building in like 30 minutes https://github.com/kd-11/rpcs3/runs/13593717063 Disables some NVIDIA speedhacks entirely. They're still enabled on pre-pascal GPUs and are known to cause crashes on RTX series.

Sure! Will let you know if anything changes.

**Comment by kd-11** [comment]
That's ok, I expected it to given the open bug reports. Unfortunately it would be extremely tedious to debug this with markers unless I have affected hardware to plug directly into my dev environment. I'll have to just shelve this for the time being.

**Comment by Valtekken** [comment]
> That's ok, I expected it to given the open bug reports. Unfortunately it would be extremely tedious to debug this with markers unless I have affected hardware to plug directly into my dev environment. I'll have to just shelve this for the time being.

Alright, so do I open the new issue or just leave everything as is?

**Comment by kd-11** [comment]
Open the ticket for the freeze.

**Comment by Valtekken** [comment]
> Open the ticket for the freeze.

Alright, done.

**Comment by kd-11** [comment]
Try with https://github.com/RPCS3/rpcs3/pull/13933. It fixed a device lost for me in "into the nexus".

**Comment by Valtekken** [comment]
> Try with #13933. It fixed a device lost for me in "into the nexus".

I tried it and it still crashed, but without any DEVICE_LOST so this is probably the other issue, #13888 , causing it

**Comment by Valtekken** [comment]
By the way, I just bought a new GPU with 8GB of VRAM (AMD this time) and I'll be testing the game as soon as I install it and the relative drivers. This way we'll at least be able to see if a different manufacturer and a higher amount of VRAM makes any difference.

**Comment by hemang2005** [comment]
> By the way, I just bought a new GPU with 8GB of VRAM (AMD this time) and I'll be testing the game as soon as I install it and the relative drivers. This way we'll at least be able to see if a different manufacturer and a higher amount of VRAM makes any difference.
 any updates?

**Comment by hemang2005** [comment]
any update?

**Comment by Valtekken** [comment]
> > By the way, I just bought a new GPU with 8GB of VRAM (AMD this time) and I'll be testing the game as soon as I install it and the relative drivers. This way we'll at least be able to see if a different manufacturer and a higher amount of VRAM makes any difference.
> 
> any updates?

It has been updated in the correct ticket since this one is apparently not the cause for my crashes/freezes anymore - #13888

**Comment by Asinin3** [comment]
Test with https://github.com/RPCS3/rpcs3/pull/14657 (if its merged when you're reading this, just test latest RPCS3 build instead)

**Comment by Jonathan44062** [comment]
still an issue, I got it as well
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/14325102/RPCS3.log)

**Comment by koksoni** [comment]
Game crushes time to time which is very annoying, but runs smooth with i5 12400 & 1080ti

**Comment by Rhaston** [comment]
[RPCS3.log](https://github.com/user-attachments/files/18171198/RPCS3.log)
I'm experiencing the same issue, very frequent crashes. I played with settings per some threads I found online and there was some improvement compared to default settings but I still get a crash every 5 to 30 minutes. After seeing this thread I tried strict rendering mode ON but the game crashed again and the performance was horrible. I tried latest Nvidia drivers and latest RPCS3 release but nothing helps. Log attached.

Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz | 8 Threads | 31.95 GiB RAM | TSC: 2.594GHz | AVX+ | FMA3
Operating system: Windows, Major: 10, Minor: 0, Build: 19045, Service Pack: none, Compatibility mode: 0
'NVIDIA GeForce GTX 970M'

**Comment by kd-11** [comment]
https://github.com/RPCS3/rpcs3/pull/17088 might improve this as it solves a memory issue on the GPU timeline.

**Comment by digant73** [comment]
it seems another crash (users having old gpu such as gtx 970) possibly fixed by #17149.
@Rhaston can you also provide an update on that issue?

**Comment by digant73** [comment]
@Valtekken, @Rhaston can you retest please?

**Comment by Valtekken** [comment]
> [@Valtekken](https://github.com/Valtekken), [@Rhaston](https://github.com/Rhaston) can you retest please?

When I left this one it seemed evident the cause of my crashes wasn't this issue. For what it's worth, I've since finished the game multiple times over without crashes so I clearly don't have anymore whatever issue was causing the crashes for sure. The problem is that multiple driver updates, a GPU switch and multiple emulator updates have occurred since the ticket was opened so I have no idea WHAT exactly it was that fixed it.

**Comment by digant73** [comment]
Understood. If the issue is no more present, please close the ticket. As far as I see from your logs, the issue was due to device-lost. That issue was due to many bugs fixed here and there in the last years, even recently (2 weeks ago) after a very long troubleshooting

**Comment by Valtekken** [comment]
> Understood. If the issue is no more present, please close the ticket. As far as I see the issue was due to device-lost. That issue was due to many bugs fixed here and there in the last years, even recently (2 weeks ago) after a very long troubleshooting

I only kept it open after being told to do so because apparently the issue was still happening to other people. If it's no longer happening I can close it.

**Comment by digant73** [comment]
let main developers decide if closing or not that ticket

**Comment by Valtekken** [comment]
> let main developers decide if closing or not that ticket

Great, I guess it's kd's call to make then.

**Comment by kd-11** [comment]
Let's just close it if its not reproducible anymore. We have had multiple big fixes towards the crashes so its not unexpected.
